### PR TITLE
Persist latest messages on disk

### PR DESCRIPTION
**Pull request by itegulov**
## Overview
This PR introduces an API for both latest messages storage and DAG storage, but only implements the latest messages storage part of it. Ignore the fact that `childrenMap`, `dataLookup` etc are not actually persisted on disk. Adapting casper code the new DAG API will come in a separate PR.

Relevant specification describing both latest messages storage and DAG storage: https://rchain.atlassian.net/wiki/spaces/CORE/pages/563544093/DAG+Storage+Specification+Draft

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-945

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] If this PR adds a new feature, this PR includes tests related to this feature.
- [x] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        }
+        .map(_.toMap)
+  }
+
+  private def updateLatestMessagesFile(validator: Validator, blockHash: BlockHash): F[Unit] =
+    for {
+      latestMessagesLogOs <- latestMessagesLogOsRef.get
+      latestMessagesCrc   <- latestMessagesCrcRef.get
+      toAppend            = validator.concat(blockHash).toByteArray
+      _                   = latestMessagesCrc.update(toAppend)
```
Comment (odeac) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: these operations having arbitrary side-effects look pretty bad. 

We should refactor these to be in sync with the rest of our code

Comment (itegulov) [reply block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: True. I will create an immutable wrapper for CRC32.

File: block-storage/src/main/scala/coop/rchain/blockstorage/util/Crc32.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  def bytes: F[Array[Byte]] =
+    value.map { value =>
+      val byteBuffer = ByteBuffer.allocate(8)
+      byteBuffer.putLong(value)
+      byteBuffer.array()
+    }
+
+  def reset: F[Unit] =
+    internal.reset().pure[F]
```
Comment (goral09) [inline block-storage/src/main/scala/coop/rchain/blockstorage/util/Crc32.scala]: This should be `Sync[F].delay { … }` as `reset` is effectful. `.pure[F]` is still eager.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      Files.move(tmpCrc, latestMessagesCrcFilePath, StandardCopyOption.REPLACE_EXISTING)
+    }
+
+  private def squashLatestMessagesDataFile(): F[Unit] =
+    for {
+      latestMessages                <- latestMessagesRef.get
+      latestMessagesLogOutputStream <- latestMessagesLogOutputStreamRef.get
+      _                             = latestMessagesLogOutputStream.close()
+      tmpSquashedData               = Files.createTempFile("rchain-block-dag-file-storage-", "-squashed-data")
+      tmpSquashedCrc                = Files.createTempFile("rchain-block-dag-file-storage-", "-squashed-crc")
```
Comment (goral09) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: These effectful function calls should be wrapped in `Sync[F].delay { … }`.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+      val tmpCrc = Files.createTempFile("rchain-block-dag-file-storage-", "-crc")
+      Files.write(tmpCrc, newCrcBytes)
+      Files.move(tmpCrc, latestMessagesCrcFilePath, StandardCopyOption.REPLACE_EXISTING)
+    }
+
+  private def squashLatestMessagesDataFile(): F[Unit] =
+    for {
+      latestMessages                <- latestMessagesRef.get
+      latestMessagesLogOutputStream <- latestMessagesLogOutputStreamRef.get
+      _                             = latestMessagesLogOutputStream.close()
```
Comment (goral09) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: This is effectfull as well.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+      tmpSquashedData               = Files.createTempFile("rchain-block-dag-file-storage-", "-squashed-data")
+      tmpSquashedCrc                = Files.createTempFile("rchain-block-dag-file-storage-", "-squashed-crc")
+      dataByteBuffer                = ByteBuffer.allocate(64 * latestMessages.size)
```
Comment (goral09) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: This is also.

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+      _ = latestMessages.foreach {
+        case (validator, blockHash) =>
+          dataByteBuffer.put(validator.toByteArray)
+          dataByteBuffer.put(blockHash.toByteArray)
```
Comment (goral09) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]:
```scala
lastMessages.traverse { case (validator, blockHash) => 
  Sync[F].delay { dataByteBuffer.put(…) }
}
```

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+      }
+      _                = Files.write(tmpSquashedData, dataByteBuffer.array())
+      squashedCrc      = Crc32.empty[F]()
```
Comment (goral09) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala]: If this is a constructor then it should return `F[Crc32[F]]`.

**Pull request merged**